clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst

```diff
@@ -0,0 +1,26 @@
+.. title:: clang-tidy - readability-mark-static
+
+readability-mark-static
+=======================
+
+Detects variable and function can be marked as static.
+
+Static functions and variables are scoped to a single file. Marking functions
+and variables as static helps to better remove dead code. In addition, it gives
+the compiler more information and can help compiler make more aggressive
+optimizations.
+
```

Comment:
Mention other benefits: avoid ODR violations

Reply:
Yes it does, because you cannot get access to such variable/function via extern in separate TU.

[(2.3)](https://eel.is/c++draft/basic.link#2.3)
When a name has [internal linkage](https://eel.is/c++draft/basic.link#def:linkage,internal), the entity it denotes can be referred to by names from other scopes in the same translation unit[.](https://eel.is/c++draft/basic.link#2.3.sentence-1)

The name of an entity that belongs to a [namespace scope](https://eel.is/c++draft/basic.scope.namespace) has internal linkage if it is the name of
[(3.1)](https://eel.is/c++draft/basic.link#3.1)
a variable, variable template, function, or function template that is explicitly declared static; or

Reply:
Did some research but did not find the case I remembered (it might have variable templates though):
- anonymous namespaces do not have external linkage until C++11: https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces / https://en.cppreference.com/w/cpp/language/storage_duration#Linkage
- `static` variable templates do not have internal linkage until C++14: https://en.cppreference.com/w/cpp/language/storage_duration#Linkage

Interesting tidbit that enumerations have external linkage.

Reply:
Any of the following names declared at namespace scope have internal linkage:
- variables, variable templates(since C++14), functions, or function templates declared static;

Looks fine... Still would be nice to allow user select static or anonymous namespace for "auto fix"

Reply:
I'm not sure about automatic fixes for anonymous namespaces. For example, in my work project we keep single monolithic anonymous namespace, so function body move would be required.